rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored. If the algorithm value is
+not set (`= UNSET`), then the `precision_config` will take effect.
+
+In general the first two types in the algorithm name are `AType` and `BType` -
+the precisions that the LHS and RHS of the operation are rounded to, and the
+third one is `AccumType` - the accumulation type. These types are independent
+from the storage types of the inputs and the output (but not all combinations
+are supported). Some algorithms have other suffixes too.
+
+This is the current list of algorithms:
+
+* `UNSET`: There is no restriction on the algorithm, the precision is decided
+  based on the `precision_config`. Use this for every type combination that
+  doesn't have a specific algorithm.
+* `ANY_F8_ANY_F8_F32`: The inputs are used with the full precision of their
+  storage type (which must be an 8-bit floating point type - the same for both
+  operands). The accumulator type is F32.
+* `ANY_F8_ANY_F8_F32_FAST_ACCUM`: The inputs are used with the full precision of
+    their storage type (which must be an 8-bit floating point type - the same
+    for both operands). The accumulator type is formally F32, but the
+    intermediate results will not periodically be promoted to a higher
+    precision.
+* `F16_F16_F16`: The inputs are used with / casted to F16 precision and the
+  accumulator type is also F16.
+* `F16_F16_F32`: The inputs are used with / casted to F16 precision and the
+  accumulator type is F32.
+* `BF16_BF16_BF16`: The inputs are used with / casted to BF16 precision and the
+  accumulator type is also BF16.
+* `BF16_BF16_F32`: The inputs are used with / casted to BF16 precision and the
+  accumulator type is F32.
+* `BF16_BF16_F32_X3`: An algorithm which uses three BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision.
+* `BF16_BF16_F32_X6`: An algorithm which uses six BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision ("similar" to F32_F32_F32).
+* `TF32_TF32_F32`: The inputs are casted to TF32 precision and the accumulator
+  type is F32.
+* `TF32_TF32_F32_X3`: An algorithm which uses 3 TF32_TF32_F32 matmuls to achieve
+  a higher precision ("similar" to F32_F32_F32).
+* `F32_F32_F32`: The inputs are used with / casted to F32 precision and the
+  accumulator type is also F32.
+* `F64_F64_F64`: The inputs are used with / casted to F64 precision and the
+  accumulator type is also F64.
+
+In general, it is not guaranteed that the each algorithm is supported on each
```

Comment:
Is there any METADATA in the output that has to be plumbed through to all StableHLO ops now?

As in, if the output of `dot_general` goes into a `add`, does the `add` now need to have a precision config? 

As in, does precision config have to go through all the StableHLO ops now?

Reply:
I don't think so, the intention is that we would only set the precision for the dot (and later convolution) operations.
This would basically modify what types do we use in the tensor cores or other hw which does the matmul.
(So even if we have F32 storage type, we can cast it to TF32 or BF16 when calculating with it)

Reply:
ack thanks.